web/src/components/text-input/generic.tsx

```diff
@@ -31,74 +49,194 @@ export interface TextInputProps extends AriaTextFieldOptions {
+      setOsInstance(textAreaRender({ props, ref, setOverflowPadding }))
+    }
+
+    return () => {
+      if (
+        props.component === 'textarea' &&
+        OverlayScrollbars.valid(osInstance)
+      ) {
+        osInstance.destroy()
+      }
+    }
+  }, [])
+
+  const usePrevious = <T extends unknown>(v: T): T | undefined => {
+    const prevRef = React.useRef<T>()
+    useEffect(() => {
+      prevRef.current = v
+    })
+    return prevRef.current
+  }
+
+  if (props.component === 'textarea') {
+    const { allowResize, autoExpand, minHeight, maxHeight, rows } = props
+
+    const prevProps = usePrevious({
+      width,
+      height,
+      maxLength,
+      allowResize,
+      autoExpand,
+      minHeight,
+      maxHeight,
+      isReadOnly,
+      isDisabled,
+      defaultValue,
+      rows,
+    })
+
+    useEffect(() => {
+      if (OverlayScrollbars.valid(osInstance)) {
+        // Re-render textarea, if these two property changed, otherwise old state will be shown
+        if (shouldScrollbarUpdate(props, prevProps)) {
+          osInstance.destroy()
+
+          setOsInstance(textAreaRender({ props, ref, setOverflowPadding }))
+        }
+      }
+    })
+  }
 
   const allowedChars = maxLength - value.length
+  const isMaxLengthReached = allowedChars < 0
+  const isMinLengthReached = value.length < minLength
 
   const { labelProps, inputProps } = useTextField(
     {
       ...props,
       maxLength: undefined,
-      onChange: p => {
-        setValue(p)
-        setDirty(p.length > 0)
-        props.onChange?.(p)
+      minLength: undefined,
+      onChange: v => {
+        setValue(v)
+        setDirty(v.length > 0)
+        props.onChange?.(v)
+      },
+      onFocus: e => {
+        if (props.component === 'textarea' && props.isReadOnly) {
+          const textarea = e.target as HTMLTextAreaElement
+          textarea.setSelectionRange(0, textarea.value.length)
+        }
       },
     },
     ref,
   )
 
-  const numberValid = useMemo(() => {
-    if (props.type !== 'number') return true
+  const min = 'min' in props ? props.min : undefined
+  const max = 'max' in props ? props.max : undefined
+
+  const isNumberValid = useMemo(() => {
+    if (props.type === 'search' || props.type !== 'number') return true
     if (value.length === 0) return true
     const val = Number(value)
     if (min !== undefined && val < min) return false
     if (max !== undefined && val > max) return false
     return true
   }, [value, min, max])
 
-  const invalid = isInvalid || allowedChars < 0 || !numberValid
+  const invalid =
+    isInvalid || isMaxLengthReached || isMinLengthReached || !isNumberValid
   const customProps = { ...props, isInvalid: invalid }
   const { onFocus } = useFocusStyle(customProps)
-  const classes = useStyles(customProps)
+  let classes
+  if (props.component === 'textarea') {
+    classes = useStyles({ ...customProps, overflowPadding })
+  } else {
+    classes = useStyles(customProps)
+  }
```

Comment:
As we started to develop a text-area without a custom scrollbar, it looked totally fine. And we didn't have specific `if` conditions for the text-area. From one side I agree That there are a lot of stuff related to text-area only, but from another component contains a lot of common stuff and in case we will separate components there will be a lot of repeats. I refactored the `generic.tsx` component and decreased the quantity of specific `if` for the text-area, hope now it looks better.